thunder/core/vjp_utils.py

```diff
@@ -178,3 +178,45 @@ def bw_fn(*args, **kwargs):
     _cache[key] = fw_fn, bw_fn
 
     return fw_fn, bw_fn
+
+
+def get_saved_for_backward_tensors(trace: TraceCtx) -> tuple[TensorProxy]:
+    """
+    Given a trace, return the tensors that are saved for backward in the trace.
+
+    Args:
+        trace: The trace to extract saved tensors from.
+
+    Example:
+        >>> import thunder
+        >>> import torch
+        >>> from thunder.core.vjp_utils import get_saved_for_backward_tensors
+        >>> def forward(a, b): return a * b
+        >>> a = torch.tensor(1.0, requires_grad=True)
+        >>> b = torch.tensor(2.0, requires_grad=True)
+        >>> jitted_forward = thunder.jit(forward)
+        >>> jitted_forward(a, b)
+        tensor(2., grad_fn=<ThunderFunctionBackward>)
+        >>> trace = thunder.last_traces(jitted_forward)[-1]
+        >>> saved_tensors = get_saved_for_backward_tensors(trace)
+        >>> print(saved_tensors)
+        (<TensorProxy(name="a", dtype=thunder.dtypes.float32, shape=())>, <TensorProxy(name="b", dtype=thunder.dtypes.float32, shape=())>)
+    """
+    # First let's check if the trace was generated by Thunder's automatic differentiation
+    # The name is defined at
+    # https://github.com/Lightning-AI/lightning-thunder/blob/681aef744786c85f925a590e6b5e6b5162932adf/thunder/core/transforms.py#L3643
+    utils.check(
+        lambda: trace.fn.__name__ == "augmented_forward_fn",
+        "The trace must be generated by Thunder's automatic differentiation",
+    )
```

Comment:
Maybe this check is not enough, while testing I ran into the case where `trace.fn` is `None`. Such behavior skips this check and fails on the `trace.output` call with a `NoneType` error. (Likely because the tensors do not require grad in my test)

repro:
```python
import thunder
import torch
from thunder.core.vjp_utils import get_saved_for_backward_tensors
def forward(a, b): return a*b
a = torch.randn((1, 5), device="cuda")
b = torch.randn((1, 5), device="cuda")
jf = thunder.jit(forward)
jf(a, b)
trace = thunder.last_traces(jf)[-1]
get_saved_for_backward_tensors(trace)
```
Outputs:
```bash
Traceback (most recent call last):
  File "/opt/lightning-thunder/test.py", line 10, in <module>
    get_saved_for_backward_tensors(trace)
  File "/opt/lightning-thunder/thunder/core/vjp_utils.py", line 218, in get_saved_for_backward_tensors
    saved_tensors = trace.output[1][0]
  File "/opt/lightning-thunder/thunder/core/proxies.py", line 1513, in __getitem__
    return method(self, key)
TypeError: 'NoneType' object is not callable
```

Reply:
Thanks, I added a test that checks that the error is thrown.